doc/widget_server.md

```diff
@@ -0,0 +1,417 @@
+
+- `"t"` is the tag of the element.
+- `"a"` are the non-event attributes of the element.
+- `"e"` are the event handlers that should be attached to the element.
+- `"c"` are the children of the element.
+- `"tt"` is an additional tooltip element to be discussed below.
+
+For example:
+
+```json
+{
+  "t": "button",
+  "a": {
+      "style": {
+          "color": "red",
+          "padding": "5em",
+          "textAlign": "left"
+      },
+      "className": "b--blue ba",
+  },
+  "e": {
+      "onClick": { "h": 0, "r": [ 51 ] }
+  },
+  "c": ["hello", "world"]
+}
+```
+
+Converts to the React element:
+
+```tsx
+<button
+    style={{color:"red", padding:"5em", textAlign:"left"}}
+    className="b--blue ba"
+    onClick={() => handle({h:0, r:[51]}, 'onClick')}>
+  helloworld
+</button>
+```
+
+
+#### Attributes
+The attributes field of `WidgetElement` takes the form of a JSON object of keys and values.
+Either both the key and value are a string, or a special case is `"style"`, whose value is a JSON object of string key/value pairs.
+Events attached to the element are placed in their own `e` field.
+Note: the keys of attributes and styles follow the React convention; for example `"textAlign"` and `"onClick"` instead of `"text-align"` and `"on-click"`. Additionally, `"className"` is used instead of `"class"`. So if you are not implementing with React, you will need to convert all of these keynames back to their HTML-native forms.
+
+#### `WidgetElement` Tooltips
+
+Widgets provide special support for 'tooltips' which are pieces of html that appear in 'popovers'. If the `tt` property is set on a `WidgetElement` then the `WidgetHTML` value of `tt` should be rendered in a popover-like element (always, not only when the user hovers their cursor over the element). In the vscode implementation, this is implemented with the popper.js library, where the given WidgetElement is wrapped in an additional `<div/>` also containing the tooltip content. See the [vscode implementation](https://github.com/leanprover/vscode-lean/blob/master/infoview/widget.tsx) for more details.
+
+### Rendering `WidgetComponent`, `null` and `string`
+
+A `WidgetHTML` with value `null` should return an empty html object (in React you can do this by returning `null` or `false`).
+If the `WidgetHTML` is a string, then return that string.
+
+A `WidgetComponent` only has one child called `c`. To render a component, simply render all of the child elements and then return them as a list.
+
+#### Examples:
+```json
+null ↦ null
+"hello" ↦ "hello"
+{ "c": ["hello", " ", "world"]} ↦ <> hello world </>
+```
+
+### Stylesheet
+
+The widgets provided in core lean assume that a CSS stylesheet called `tachyons.css` is loaded. This can be downloaded from tachyons.io. Note that this follows the 'functional CSS' paradigm, so the idea is that the Lean widget-writer should never need to write their own styles, and instead attaches these tachyons classes to their elements.
```

Comment:
```suggestion
The widgets provided in core Lean assume that a CSS stylesheet called `tachyons.css` is loaded. This can be downloaded from <https://tachyons.io>. Note that this follows the 'functional CSS' paradigm, so the idea is that the Lean widget-writer should never need to write their own styles, and instead attaches these tachyons classes to their elements.
```